src/dag/chunk.rs

```diff
@@ -1,8 +1,100 @@
-pub type Key = String;
+use super::meta_generated;
+use flatbuffers;
+use flatbuffers::FlatBufferBuilder;
 
 #[allow(dead_code)]
+#[derive(Debug)]
 pub struct Chunk {
-    pub key: Key,
-    buf: Vec<u8>,
-    refs: [Key],
+    hash: String,
+    data: Vec<u8>,
+    meta: Option<(Vec<u8>, usize)>,
+}
+
+#[allow(dead_code)]
+impl Chunk {
+    pub fn new(hash: String, data: Vec<u8>, refs: &[&str]) -> Chunk {
+        Chunk {
+            hash,
+            data,
+            meta: Chunk::create_meta(refs),
+        }
+    }
+
+    pub fn read(hash: String, data: Vec<u8>, meta: Option<Vec<u8>>) -> Chunk {
+        Chunk {
+            hash,
+            data,
+            meta: meta.map(|v| (v, 0)),
+        }
+    }
+
+    pub fn hash(&self) -> &str {
+        &self.hash
+    }
+
+    pub fn data(&self) -> &[u8] {
+        &self.data
+    }
+
+    pub fn refs<'a>(&'a self) -> Box<dyn Iterator<Item = &str> + 'a> {
```

Comment:
I was very sad to return the boxed iterator, but I wanted to flatten then case of no meta buffer (as will be the case for leaf nodes of the dag) to an empty iterator so that callers don't need to worry about. In that case, the return iterator can either be the generated one from FlatBuffers or std::iter::Empty.

I couldn't find any other way to flatten these cases... If the FlatBuffers had a way to get out &[&str] that would work, but I don't think it does.

Reply:
Yeah, I agree that this is annoying; I'd at least like to avoid the alloc for the empty case, but another problem is that the caller is taking ownership of whatever you return, and so you can't just return a static lifetime empty iterator there.

I'd just leave another TODO for now. Another way I would try attacking this later is that if we only expect callers to call this if they plan to consume the whole ref list, you can have them pass in a `mut Vec<&str>`, and copy any refs into it after clearing it, allowing the Vector to be on the caller's stack frame.

Reply:
Sure but then I might as well just return a new `Vec<String>` or whatever, right? What is flatbuffers really buying us over pb at that point?

Reply:
> another problem is that the caller is taking ownership of whatever you return,

I don't understand this comment. I tried to make it so that callers *is not* taking ownership of returned data (flatbuffer still owns it).

Reply:
I just decided to have it return an option instead. Still added TODO.